### ISSUE TITLE
[camel 3.13] NativeImageResourceBuildItem fails for Atlasmap extension

### ISSUE THREAD

**Issue by ffang** [apache/camel-quarkus#3189]
get error like
```
resteasy, resteasy-jackson, smallrye-context-propagation, vertx]
2021-10-15 12:30:49,017 ERROR [io.atl.cor.DefaultAtlasContext] (executor-thread-0) Cannot find module info for the DataSource uri 'atlas:csv:target-csv?firstRecordAsHeader=true'
2021-10-15 12:30:49,017 ERROR [io.atl.cor.DefaultAtlasContext] (executor-thread-0) Cannot find module info for the DataSource uri 'atlas:json:Contact'
2021-10-15 12:30:49,019 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /atlasmap/json/csv2json failed, error id: 44dfe6b5-6bd3-4e8e-a6f2-58b52d54d041-1: org.jboss.resteasy.spi.UnhandledException: org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[]
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:135)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:90)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:829)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:567)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[]
	at org.apache.camel.support.ExchangeHelper.extractResultBody(ExchangeHelper.java:677)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.extractResultBody(DefaultProducerTemplate.java:591)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.extractResultBody(DefaultProducerTemplate.java:587)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.requestBody(DefaultProducerTemplate.java:414)
	at org.apache.camel.quarkus.component.atlasmap.it.AtlasmapResource.convertCsv2JsonWithJson(AtlasmapResource.java:120)
	at org.apache.camel.quarkus.component.atlasmap.it.AtlasmapResource_ClientProxy.convertCsv2JsonWithJson(AtlasmapResource_ClientProxy.zig:219)
	at java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	... 17 more
Caused by: io.atlasmap.api.AtlasException: Errors: [Module not found for docId 'target-csv': Document='null(ID:target-csv)', path='/<>/firstName'], [Module not found for field type='STRING', path='/firstName': Document='null(ID:Contact)', path='/firstName'], [Module not found for docId 'target-csv': Document='null(ID:target-csv)', path='/<>/lastName'], [Module not found for field type='STRING', path='/lastName': Document='null(ID:Contact)', path='/lastName'], [Module not found for docId 'target-csv': Document='null(ID:target-csv)', path='/<>/age'], [Module not found for field type='STRING', path='/age': Document='null(ID:Contact)', path='/age'], 
	at org.apache.camel.component.atlasmap.AtlasMapEndpoint.onExchange(AtlasMapEndpoint.java:198)
	at org.apache.camel.support.ProcessorEndpoint$1.process(ProcessorEndpoint.java:61)
	at org.apache.camel.support.AsyncProcessorConverterHelper$ProcessorToAsyncProcessorBridge.process(AsyncProcessorConverterHelper.java:66)
	at org.apache.camel.impl.engine.SharedCamelInternalProcessor.process(SharedCamelInternalProcessor.java:217)
	at org.apache.camel.impl.engine.SharedCamelInternalProcessor$1.process(SharedCamelInternalProcessor.java:111)
	at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.process(DefaultAsyncProcessorAwaitManager.java:83)
	at org.apache.camel.impl.engine.SharedCamelInternalProcessor.process(SharedCamelInternalProcessor.java:108)
	at org.apache.camel.support.cache.DefaultProducerCache.send(DefaultProducerCache.java:190)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.send(DefaultProducerTemplate.java:176)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.send(DefaultProducerTemplate.java:172)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.requestBody(DefaultProducerTemplate.java:413)
	... 31 more

2021-10-15 12:30:49,431 ERROR [io.atl.cor.DefaultAtlasContext] (executor-thread-0) Cannot find module info for the DataSource uri 'atlas:xml:-LO2lU-dzmk816-JEluS'
2021-10-15 12:30:49,432 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /atlasmap/json/java2xml failed, error id: 44dfe6b5-6bd3-4e8e-a6f2-58b52d54d041-2: org.jboss.resteasy.spi.UnhandledException: org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[]
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:135)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:90)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:829)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:567)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[]
	at org.apache.camel.support.ExchangeHelper.extractResultBody(ExchangeHelper.java:677)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.extractResultBody(DefaultProducerTemplate.java:591)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.extractResultBody(DefaultProducerTemplate.java:587)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.requestBody(DefaultProducerTemplate.java:414)
	at org.apache.camel.quarkus.component.atlasmap.it.AtlasmapResource.convertJava2XmlWithJson(AtlasmapResource.java:90)
	at org.apache.camel.quarkus.component.atlasmap.it.AtlasmapResource_ClientProxy.convertJava2XmlWithJson(AtlasmapResource_ClientProxy.zig:188)
	at java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	... 17 more
Caused by: io.atlasmap.api.AtlasException: Errors: [Module not found for field type='STRING', path='/tns:Person/tns:firstName/@value': Document='null(ID:-LO2lU-dzmk816-JEluS)', path='/tns:Person/tns:firstName/@value'], [Module not found for field type='STRING', path='/tns:Person/tns:lastName/@value': Document='null(ID:-LO2lU-dzmk816-JEluS)', path='/tns:Person/tns:lastName/@value'], [Module not found for field type='INTEGER', path='/tns:Person/tns:age/@value': Document='null(ID:-LO2lU-dzmk816-JEluS)', path='/tns:Person/tns:age/@value'], 
	at org.apache.camel.component.atlasmap.AtlasMapEndpoint.onExchange(AtlasMapEndpoint.java:198)
	at org.apache.camel.support.ProcessorEndpoint$1.process(ProcessorEndpoint.java:61)
	at org.apache.camel.support.AsyncProcessorConverterHelper$ProcessorToAsyncProcessorBridge.process(AsyncProcessorConverterHelper.java:66)
	at org.apache.camel.impl.engine.SharedCamelInternalProcessor.process(SharedCamelInternalProcessor.java:217)
	at org.apache.camel.impl.engine.SharedCamelInternalProcessor$1.process(SharedCamelInternalProcessor.java:111)
	at org.apache.camel.impl.engine.DefaultAsyncProcessorAwaitManager.process(DefaultAsyncProcessorAwaitManager.java:83)
	at org.apache.camel.impl.engine.SharedCamelInternalProcessor.process(SharedCamelInternalProcessor.java:108)
	at org.apache.camel.support.cache.DefaultProducerCache.send(DefaultProducerCache.java:190)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.send(DefaultProducerTemplate.java:176)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.send(DefaultProducerTemplate.java:172)
	at org.apache.camel.impl.engine.DefaultProducerTemplate.requestBody(DefaultProducerTemplate.java:413)
	... 31 more

2021-10-15 12:30:49,454 ERROR [io.atl.cor.DefaultAtlasContext] (executor-thread-0) Cannot find module info for the DataSource uri 'atlas:xml:Patient-3b09a669-0fa6-48e8-8220-957def0eb00f'
2021-10-15 12:30:49,454 ERROR [io.atl.cor.DefaultAtlasContext] (executor-thread-0) Cannot find module info for the DataSource uri 'atlas:json:Person_3-3da3e964-ba87-4443-a2ec-703219af5cb4'
2021-10-15 12:30:49,455 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /atlasmap/adm/xml2json failed, error id: 44dfe6b5-6bd3-4e8e-a6f2-58b52d54d041-3: org.jboss.resteasy.spi.UnhandledException: org.apache.camel.CamelExecutionException: Exception occurred during execution on the exchange: Exchange[]
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:135)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:90)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:829)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:567)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)

```

**Comment by ffang** [comment]
I think output like
```
2021-10-15 12:30:49,017 ERROR [io.atl.cor.DefaultAtlasContext] (executor-thread-0) Cannot find module info for the DataSource uri 'atlas:csv:target-csv?firstRecordAsHeader=true'
2021-10-15 12:30:49,017 ERROR [io.atl.cor.DefaultAtlasContext] (executor-thread-0) Cannot find module info for the DataSource uri 'atlas:json:Contact'
```
is suspicious.

And I'm pretty sure this is caused by the atlasmap upgrade from 2.2.3 to 2.3.x recently.

I rollback too use atlasmap 2.2.3 in camel-atlasmap component and build it, then build the atlasmap camel-quarkus extension, and the integration test passed in native mode

**Comment by zbendhiba** [comment]
Yes I'm on this one

**Comment by ffang** [comment]
Thanks @zbendhiba  for picking this up!

**Comment by zbendhiba** [comment]
I was already investigating. I do not necessary open issues for camel-main.

**Comment by zbendhiba** [comment]
**Current status :**

Not done with this one. At the first step, I looked at ClassLoader, but notthing war broken there.
I had to add some conf locally to get some logs.
Since atlasmap 2.3.x, there is only one of the modules that is loaded. This line of code https://github.com/atlasmap/atlasmap/blob/883cff34cdce946d3cf097a01f1f188a2cd51dcf/lib/core/src/main/java/io/atlasmap/core/DefaultAtlasContextFactory.java#L324 finds only one single module compared to previous version. (on native mode only)


So since upgrade of this version of atlasmap, this camel-quarkus configuration does not work anymore :
https://github.com/apache/camel-quarkus/blob/64e573611bfb67c83b0603709d104b5470949906/extensions/atlasmap/deployment/src/main/java/org/apache/camel/quarkus/component/atlasmap/deployment/AtlasmapProcessor.java#L78

Itseems that it takes only the first META-INF/services/atlas/module/atlas.module file. Meaning I had at first shot only the Java module loaded, then I exluded this one from the extension, then it has loaded just the Json module ...

I'm still investigating on what could have generated this. Putting this here, in case someone has found this bug before.

**Comment by jamesnetherton** [comment]
I disabled the native itests so that we can get a successfull build of `camel-main`.

**Comment by zbendhiba** [comment]
Finally found out why it fails. Monday, I'll propose a Substitution for next release and see how we could fix this in Atlasmap. I'll address also the changes needed to be done for  this release : jandex for instance.
This is [the problematic line](https://github.com/atlasmap/atlasmap/blob/15d9e76f743ce67438316e28bf3c7d6a79a50513/lib/core/src/main/java/io/atlasmap/core/DefaultAtlasCompoundClassLoader.java#L68). 

In this [commit](https://github.com/atlasmap/atlasmap/commit/04254c4a611213152f7e537c1b4e1da34127b746), the LinkedList has been replaced by a LinkedHashSet. This works fine on JVM mode, as the different URLs loaded have different names and paths. And if we debug, we can see different URLs from the different modules : JAVA, JSON, CSV, XML

In other hands, on native image, NativeImageResourceBuildItem, we get same name & path of URL for all the files in the different atlasmap modules. Always beginning from META-INF, so same URL. 
I'm wondering if there is some issue to address in Quarkus. WDYT @ppalaga @jamesnetherton ? 

Anyway, it's kind of difficult to debug. As the name, path ... is the same. But with LinkedList, it loads later the right 4 modules. So I would love to dig deeper, but I would need some help or have more time to investigate.

**Comment by zhfeng** [comment]
@zbendhiba I think it needs to raise an issue on Quarkus to get their feedback ? and if you can narrow down it out side of Atlasmap, e.g. a reproducer only use ```LinkedHashSet```, that could be very helpful for investigating.

**Comment by ppalaga** [comment]
> In other hands, on native image, NativeImageResourceBuildItem, we get same name & path of URL for all the files in the different atlasmap modules. Always beginning from META-INF, so same URL.

I'd say it works as designed in native mode and it is hard to even conceive that it could work differently. Just think of it: what is the native executable produced by `native-image`? It is a BLOB containing a minimal VM plus the whole application code. There are no separate jars there. So the classpath has to be flat and all resources have to be accessible through a single classloader.

**Comment by ppalaga** [comment]
It looks like AtlasMap is making an assumption about `java.lang.ClassLoader.getResources(String)` that does not hold for the native mode. Possible fix strategies that come to my mind:

* Compute URL equality based on content returned via URL.openStream() instead of URL.hashCode() & URL.equals()
* Check whether some Identity Set implementation would work as expected in both JVM and native mode. Just note the impl would have to keep ordering to be consistent with `java.lang.ClassLoader.getResource(String)` which `Collections.newSetFromMap(new IdentityHashMap())` does not.

**Comment by zbendhiba** [comment]
Well, it would be easy if we can easily change the Classloader too. As we won't have this list of possible Classloaders..
I tried to do it using this work done here : [atlasmap-2704](https://github.com/atlasmap/atlasmap/issues/2704,) but it is still complex. There are different places, and ways to create the classloader, the context ... So I fixed at some point some tests, and had others still failing. So I abandoned this, as I'm not going to substitute everywhere. We need a different approach in AtlasMap, and maybe even some changes in the Camel Component itself, on how we setup the context.
